[클린코드 2기 장태영] 계산기 미션

taeyoungs opened the pull request:
안녕하세요. 클린코드 2기 장태영입니다.
잘 부탁드립니다! 🙇‍♂️

조금만 더 ! 조금만 더 ! 하다가 처음에 생각했던 때보다 늦게 PR을 올리게 됐네요. 😅

![스크린샷 2022-03-12 오후 6 37 20](https://user-images.githubusercontent.com/52184469/158013253-87c8db8d-df08-4b73-b98e-965ee52e5db5.png)
위와 같은 구조를 손으로 미리 그린 다음에 작업을 하긴 했는데 의도대로 잘 구현이 됐는지 모르겠습니다.
~~테스트 요구사항은 아직 구현하지 못했는데 오늘하고 내일 중으로 작업해서 올려보도록 하겠습니다!~~

(2022.03.13 Update) Cypress 테스트 코드 추가 완료했습니다!

[데모 링크](https://taeyoungs-js-calculator.netlify.app/)

## 🎯  기능 요구사항
- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
- [x] 2개의 숫자에 대해 곱셈이 가능하다.
- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.

## 📊 테스트 요구사항
- [x] 기능 요구사항에 제시된 7개의 항목에 대해 테스트 케이스를 만든다.

## ❓ 구현하면서 드는 생각
1. 다른 분들 코드를 천천히 살펴보니 이런 구조가 굳이 필요했나 싶기도 하네요. 😥
2. 실패 케이스에 대한 테스트 코드도 작성하고 싶은데 감이 잘 오지 않네요. 실패한다고 가정하고 결과를 그렇게 작성하면 될까요 .. ?

taeyoungs committed:
chore: ESLint, Prettier 설정

.eslintrc
- ESLint, Prettier Recommend rule 적용

.prettierrc
- 세미콜론, 홑따옴표 true 적용
- tabWidth를 4로 적용

taeyoungs committed:
mod: index.html에 script 태그 추가

taeyoungs committed:
add: common 파일 작성(utils, constants, validations)

common/constants.js
- 프로젝트 내에서 사용되는 고정 메시지 작성

common/utils.js
- 피연산자와 연산자가 포함된 data 객체를 파라미터로 받아 계산된
  결과를 반환해주는 유틸 함수 구현

common/validations.js
- 피연산자와 연산자 클릭 시 호출할 검증 함수 구현

taeyoungs committed:
add: Calculator 컴포넌트 내에 위치할 컴포넌트 구현

componentns/Digits.js
- 숫자 버튼 상위 컴포넌트 구현

components/Modifiers.js
- AC 버튼 상위 컴포넌트 구현

components/Operations.js
- 연산자 버튼 상위 컴포넌트 구현

taeyoungs committed:
add: Calculator 컴포넌트 구현 및 index.js 수정

components/Calculator.js
- Digits, Modifiers, Operations를 하위 컴포넌트로 갖는 Calculator
  컴포넌트 구현

index.js
- Calculator 컴포넌트 initialState 작성
- Calculator 생성자 함수 호출 구문 추가

taeyoungs committed:
chore: Cypress initialize

taeyoungs committed:
add: js-calculator Cypress 테스트 코드 작성

index.html
- Cypress 테스트 코드에서 사용할 data-cy 속성을 각 태그에 추가

caculator.spec.js
- Github 테스트 요구사항을 포함한 Cypress 테스트 코드 작성

JunilHwang commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,107 @@
+    });
+
+    it('두 피연산자에 대해 덧셈이 가능한가?', () => {
+        cy.get('[data-cy=btn-digit-5]').click();
```
```js
const handleClickNumber = (number) => cy.get(`[data-cy=btn-digit-${number}]`).click();

handleClickNumber(5);
```
이렇게 함수로 만들어서 표현할 수도 있답니다!

그리고 data-cy는 왠지 cypress 때문에 지정한 속성 같은데..
테스트 코드를 작성하기 위해서 어플리케이션의 코드를 수정 하는 것은 지양해야 한답니다!
현재 어플리케이션 코드가 테스트하기 좋은 구조인가? 에 대해 먼저 생각해보고
테스트하기 적합한 구조로 만드는게 더 좋을 것 같아요 ㅎㅎ
cypress는 본문의 태그를 선택하는 다양한 방법이 있기 때문에 꼭 get으로 한 번에 선택하지 않아도 된답니다!
공식문서를 조금 더 차근차근 살펴보세요!

taeyoungs replied:
`테스트 코드를 작성하기 위해서 어플리케이션의 코드를 수정 하는 것은 지양해야 한답니다!` 🥲  제가 쉬운 길을 택하려고 했던 것 같네요.

공식문서를 살펴보면서 data-cy 속성들 없이 한번 진행해보겠습니다. :)

그리고 함수로 표현하는 방법이 훨씬 가독성과 재사용성이 좋네요. 반영해보겠습니다!

JunilHwang commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,107 @@
+
+        cy.on('window:alert', alertStub);
+
+        cy.get('[data-cy=btn-digit-5]').click();
+        cy.get('[data-cy=btn-digit-4]').click();
+        cy.get('[data-cy=btn-digit-3]').click();
+        cy.get('[data-cy=btn-digit-2]')
```
앞선 리뷰를 토대로 이 부분도 개선해보자면
```js
const handleClickNumbers = (...numbers) => numbers.forEach(number => cy.get(`[data-cy=btn-digit-${number}]`).click());
```
이렇게 함수를 하나 만든 다음에
```js
handleClickNumbers(5, 4, 3, 2);
```
이렇게 표현할 수 있겠네요!

taeyoungs replied:
깔끔하고 좋네요!
index.html을 이미 변경해버린 후라 변경 전이라고 가정하고 `현재 만들어진 dom을 어떻게 하면 더 정확하게 선택할 수 있을지`에 초점을 맞춰서 `caculator_with_feedback.spec.js` 이름의 테스트 코드를 추가로 작성해 봤습니다!

JunilHwang commented on index.html:
```diff
@@ -5,32 +5,33 @@
+                    <button class="digit" data-cy="btn-digit-8">8</button>
+                    <button class="digit" data-cy="btn-digit-7">7</button>
+                    <button class="digit" data-cy="btn-digit-6">6</button>
+                    <button class="digit" data-cy="btn-digit-5">5</button>
+                    <button class="digit" data-cy="btn-digit-4">4</button>
+                    <button class="digit" data-cy="btn-digit-3">3</button>
+                    <button class="digit" data-cy="btn-digit-2">2</button>
+                    <button class="digit" data-cy="btn-digit-1">1</button>
+                    <button class="digit" data-cy="btn-digit-0">0</button>
                 </div>
                 <div class="modifiers subgrid">
-                    <button class="modifier">AC</button>
+                    <button class="modifier" data-cy="btn-modifier">AC</button>
                 </div>
                 <div class="operations subgrid">
-                    <button class="operation">/</button>
-                    <button class="operation">X</button>
-                    <button class="operation">-</button>
-                    <button class="operation">+</button>
-                    <button class="operation">=</button>
+                    <button class="operation" data-cy="btn-oper-div">/</button>
+                    <button class="operation" data-cy="btn-oper-mul">X</button>
+                    <button class="operation" data-cy="btn-oper-sub">-</button>
+                    <button class="operation" data-cy="btn-oper-plus">+</button>
+                    <button class="operation" data-cy="btn-oper-cal">=</button>
```
앞선 리뷰를 참고해주세요!

JunilHwang commented on src/js/common/constants.js:
```diff
@@ -0,0 +1,4 @@
+export const OPERAND_FIRST_MESSAGE =
+    '숫자를 먼저 입력한 후 연산자를 입력해주세요!';
+export const MAX_LENGTH_MESSAGE = '숫자는 세 자리까지만 입력 가능합니다!';
+export const ONLY_TWO_OPERAND_MESSAGE = '2개의 숫자 조합만 계산 가능합니다!';
```
상수로 분리하기 👍

taeyoungs replied:
피드백으로 적극적인 상수 사용을 말씀해주셔서 기왕 더 적극적으로 사용하는 김에 기존에 작성했던 상수도 다음과 같이 바꿔봤습니다.
```javascript
export const MESSAGE = {
    OPERAND_FIRST: '숫자를 먼저 입력한 후 연산자를 입력해주세요!',
    MAX_LENGTH: '숫자는 세 자리까지만 입력 가능합니다!',
    ONLY_TWO_OPERAND: '2개의 숫자 조합만 계산 가능합니다!',
};
```

JunilHwang commented on src/js/common/utils.js:
```diff
@@ -0,0 +1,36 @@
+            break;
+        default:
+            calculatedValue = leftOperand;
+    }
```
/, X, -, + 등의 심볼도 상수로 만들어서 관리해주면 어떨까요?

taeyoungs replied:
확실히 제가 상수 사용이 익숙하지 않았던 것 같네요. 다음과 같이 만들어서 사용하는 방식으로 변경했습니다!
```javascript
export const OPERATIONS = {
    PLUS: '+',
    MINUS: '-',
    DIVIDE: '/',
    MULTIPLE: 'X',
    CALCULATE: '=',
};
```

JunilHwang commented on src/js/common/utils.js:
```diff
@@ -0,0 +1,36 @@
+            break;
+        default:
+            calculatedValue = leftOperand;
+    }
```
switch보다 더 간단하게 표현할 수 있는 방법도 있지 않을까요?

taeyoungs replied:
다른 분들의 PR 리뷰 내용을 보니 Lookup table을 추천하셔서 해당 방법을 사용해 봤습니다!
```javascript
export const CALCULATE = {
    [OPERATIONS.PLUS]: (a, b) => +a + +b,
    [OPERATIONS.MINUS]: (a, b) => +a - +b,
    [OPERATIONS.MULTIPLE]: (a, b) => +a * +b,
    [OPERATIONS.DIVIDE]: (a, b) => Math.floor(+a / +b),
};

export function calculate(data) {
    const { leftOperand, operation } = data;
    const rightOperand = data.rightOperand ? data.rightOperand : leftOperand;

    return operation
        ? CALCULATE[operation](leftOperand, rightOperand)
        : leftOperand;
}
```

JunilHwang commented on src/js/common/validations.js:
```diff
@@ -0,0 +1,41 @@
+        alert(MAX_LENGTH_MESSAGE);
+        return true;
+    }
+    return false;
```
```suggestion
    const invalid = num.length >= 3;
    if (invalid) {
        alert(MAX_LENGTH_MESSAGE);
    }
    return invalid;
```
이렇게 표현할 수 있답니다!

taeyoungs replied:
작은 차이인데 보기 훨씬 좋아지네요. 👍 
앞으로도 이런 식으로 한번 작성해 봐야겠습니다. :)

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,10 @@
+    operation: null,
+};
+
+new Calculator({ initialState });
```
state를 외부에서 주입해줄 필요가 있을까요?
Calculator 내부에서 관리해도 무방할 것 같아요!

taeyoungs replied:
음 .. 굳이 ? 라는 생각이 듭니다. Calculator 내부에서 관리하도록 변경하겠습니다!

JunilHwang commented on src/js/common/validations.js:
```diff
@@ -0,0 +1,41 @@
+        return true;
+    }
+    return false;
+}
```
```suggestion
/**
 * @param {string} operand
 * @returns {boolean} operand이 '0'이면 true 아니라면 false
 */
export function isOperandFirst(operand) {
    if (operand === '0') {
        alert(OPERAND_FIRST_MESSAGE);
    }
    return operand === '0';
}

/**
 * @param {string | null} operand
 * @returns {boolean} operand가 null이 아니라면 true null이라면 false
 */
export function isOperandFull(operand) {
    if (operand) {
        alert(ONLY_TWO_OPERAND_MESSAGE);
    }
    return Boolean(operand);
}
```
이런식으로 표현할 수 있겠네요!

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,119 @@
+ * @param {IProps} props
+ */
+function Calculator({ initialState }) {
+    this.state = initialState;
```
앞서 이야기드린 것 처럼 state는 내부에서 정의해서 가지고 있는게 더 관리하기 쉬울 것 같습니다.

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,119 @@
+/**
+ * @param {IProps} props
+ */
+function Calculator({ initialState }) {
```
아예 class로 정의해보면 어떨까요?

taeyoungs replied:
준일님 말씀대로 class로 바꿔봤습니다!

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,119 @@
+        const { operation } = this.state;
+        let { leftOperand, rightOperand } = this.state;
+
+        if (!operation) {
+            if (isMaxLength(leftOperand)) return;
+
+            if (leftOperand === '0') {
+                leftOperand = digit;
+            } else {
+                leftOperand = leftOperand + digit;
+            }
+        } else {
+            if (rightOperand && isMaxLength(rightOperand)) return;
+
+            if (rightOperand === '0' || !rightOperand) {
+                rightOperand = digit;
+            } else {
+                rightOperand = rightOperand + digit;
+            }
+        }
```
if가 두 겹으로 되어있고, else 구문도 있네요 ㅠㅠ
이런 부분은 함수로 분리해서 관리해주면 더 좋을 것 같아요!

taeyoungs replied:
함수로 분리하니 세상 편하네요. 진작에 이 부분도 분리해서 작성할 생각을 못한 게 아쉽네요.
준일님이 아래에 적어주신 코드를 반영해서 다음과 같이 분리해 봤습니다.
```javascript
export function formatOperand(operand, digit) {
    if (!operand) return digit;

    if (isMaxLength(operand, 3)) return operand;

    return `${Number(operand + digit)}`;
}
```

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,119 @@
+        if (!operation) {
+            if (isMaxLength(leftOperand)) return;
+
+            if (leftOperand === '0') {
+                leftOperand = digit;
+            } else {
+                leftOperand = leftOperand + digit;
+            }
```
```suggestion
            leftOperand = `${Number(leftOperand + digit)}`;
```
이렇게 표현할 수 있답니다!

JunilHwang commented on src/js/common/validations.js:
```diff
@@ -0,0 +1,41 @@
+ * @returns {boolean} 길이가 3 이상 이라면 true 2 이하라면 false
+ */
+export function isMaxLength(num) {
+    if (num.length >= 3) {
```
isMaxLength가 3을 내재하고 있는 상태입니다.
MaxLength를 매개변수로 주입할 수 있다면 더 유연해지지 않을까요?

taeyoungs replied:
공감합니다. 파라미터를 하나 더 주는 것만으로도 함수가 조금 더 유연해질 수 있겠네요. 👍  반영해보겠습니다.

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,119 @@
+        });
+    };
+
+    new Digits({ $target: this.$digits, updateOperand: this.updateOperand });
+
+    this.resetState = () => {
+        this.setState(initialState);
+    };
+
+    new Modifiers({ $target: this.$modifiers, resetState: this.resetState });
```
컴포넌트를 만들어주는 함수를 하나 만들어서 관리해주면 어떨까요?
지금은 코드의 흐름이 무척 어색한 것 같아요!

taeyoungs replied:
class로 변경하면서 initializeComponents라는 메서드를 추가하고 생성자 내부에서 호출하는 방식으로 변경했습니다.
좋은 방법인지는 모르겠지만 이렇게 하면 코드의 흐름이 더 좋아진걸까요 ..? 🤔

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,119 @@
+
+    new Modifiers({ $target: this.$modifiers, resetState: this.resetState });
+
+    this.setOperation = (operation) => {
+        let total;
+
+        if (isOperandFirst(this.state.leftOperand)) return;
+
+        if (operation === '=') {
+            total = calculate(this.state);
+            this.setState({
+                ...this.state,
+                total,
+                leftOperand: total,
+                rightOperand: null,
+                operation: null,
+            });
+            return;
+        }
+
+        if (isOperandFull(this.state.rightOperand)) return;
+
+        total = this.state.leftOperand + operation;
+        this.setState({
+            ...this.state,
+            total,
+            operation,
+        });
+    };
```
전체적으로 `this.state` 가 많이 사용되고 있습니다.
```js
const { state } = this;
```
처럼 state를 추출하여 사용하면 더 깔끔해질 것 같아요!

JunilHwang commented on src/js/components/Digits.js:
```diff
@@ -0,0 +1,27 @@
+ * @param {IProps} props
+ */
+function Digits({ $target, updateOperand }) {
+    this.onClickDigit = (event) => {
```
보통 이럴 땐 `handleClickDigit` 처럼 네이밍 한답니다!

JunilHwang reviewed (changes requested):
안녕하세요 태영님! 계산기 미션을 함께하게 된 리뷰어 황준일입니다 🙇‍♂️
리뷰가 많이 늦어져서 죄송합니다 🥲

리뷰의 내용을 전체적으로 정리해보자면
- 상수를 더 적극적으로 사용해보시면 좋을 것 같습니다
- 변수를 어디에 선언하냐에 따라 코드의 흐름을 이해하는 과정이 많이 달라집니다. 만약에 코드에 변경이 생긴다고 가정했을 때 이게 직관적으로 수정하기 편한 구조인지 지속적으로 고민해보시면 도움이 된답니다!
- es6의 class 문법을 이용해보면 더 깔끔하게 코드를 작성할 수 있답니다!
- if가 중첩되지 않도록 신경써주세요! 어떻게 하면 if가 중첩되지 않도록 코드를 작성할 수 있을지 고민해보시면 좋을 것 같습니다 ㅎㅎ
  - if가 중첩된다는 것은 코드의 갈래가 많아진다는 이야기이고, 코드의 갈래가 많아질수록 다른 사람 입장에서는 코드를 읽기가 점점 어려워질 수 있습니다. 그래서 if/while/for 등으로 중첩되는 코드들이 없도록 신경써서 작성하면 좋답니다!
- cypress에서 테스트를 위해 dom을 직접적으로 건드리는 방법보단, 오히려 현재 만들어진 dom을 어떻게 하면 더 정확하게 선택할 수 있을지 고민해보시면 좋을 것 같습니다. 오직 테스트를 목적으로 어플리케이션 코드를 고치는 것은 목적이 뒤바뀐 것이라고 할 수 있어요! 현재 코드로도 테스트가 가능하도록 만들어야 한답니다!

더 자세한 내용은 리뷰를 확인해주세요! 감사합니다 😁

taeyoungs committed:
mod: 더 적극적인 상수 사용 피드백 반영

taeyoungs committed:
mod: util, validation 함수 피드백 반영

utils.js
- switch 구문 대신 lookup table 방식을 이용하게끔 수정
- 복잡했던 if ~ else 구문을 함수로 분리하여 if 구문 최소화

validations.js
- isMaxLength 함수에 maxLength라는 추가 파라미터를 전달받아
  함수 자체가 조금 더 유동적으로 사용할 수 있도록 변경
- 조건문 분기값을 분리하여 조금 더 깔끔한 코드가 되도록 수정

taeyoungs committed:
mod: 이벤트 핸들러 콜백함수 네이밍 변경 (on- => handle-)

taeyoungs committed:
mod: Calculator 컴포넌트 피드백 반영

index.js
- 외부에서 주입하던 initalState 제거

Calculator.js
- state를 더 이상 내부에서 주입받지 않고 컴포넌트 내부에서 관리
  하도록 변경
- class로 전환
- 복잡했던 if ~ else 구문을 함수로 분리
- 컴포넌트들을 initialize하는 메서드 추가

taeyoungs committed:
add: Cypress test 피드백 반영 및 버그 수정

JunilHwang commented on src/js/common/constants.js:
```diff
@@ -0,0 +1,20 @@
+    CALCULATE: '=',
+};
+
+export const CALCULATE = {
+    [OPERATIONS.PLUS]: (a, b) => +a + +b,
+    [OPERATIONS.MINUS]: (a, b) => +a - +b,
+    [OPERATIONS.MULTIPLE]: (a, b) => +a * +b,
+    [OPERATIONS.DIVIDE]: (a, b) => Math.floor(+a / +b),
+};
```
```suggestion
export const CALCULATE = {
    [OPERATIONS.PLUS]: (a, b) => Number(a) + Number(b),
    [OPERATIONS.MINUS]: (a, b) => Number(a) - Number(b),
    [OPERATIONS.MULTIPLE]: (a, b) => Number(a) * Number(b),
    [OPERATIONS.DIVIDE]: (a, b) => Math.floor(Number(a) / Number(b),
};
```
이렇게 명시적으로 Number라고 표현해주면 어떨까요?

그리고 Math.floor의 경우 음수에 대해서는 생각하는 것과 다른 동작이 수행될 수 있습니다.
<img width="138" alt="image" src="https://user-images.githubusercontent.com/18749057/158188428-ecd754ec-d4f8-48cc-a6c2-375e12db9747.png">

이럴 때 `~~` 연산을 하면 아예 소숫점을 버린답니다!

참고로 `~`는 1의 보수 연산을 수행하고, 이 과정에서 모든 소숫점을 버립니다.
`~~` 처럼 1의 보수 연산을 두 번 수행하게 되면 소숫점만 버리는 연산을 수행할 수 있답니다!

<img width="129" alt="image" src="https://user-images.githubusercontent.com/18749057/158189038-8883cba1-8aac-4571-879d-cac4174255bb.png">

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,103 @@
+    operation: null,
+};
+class Calculator {
+    constructor() {
+        this.state = initialState;
+        this.$digits = document.querySelector('div.digits');
+        this.$modifiers = document.querySelector('div.modifiers');
+        this.$operations = document.querySelector('div.operations');
+        this.$total = document.querySelector('#total');
+        this.initializeComponents();
+    }
```
```suggestion
    state;
    $digits;
    $modifiers;
    $operations;
    $total;
    
    constructor() {
        this.state = { ...initialState };
        this.$digits = document.querySelector('div.digits');
        this.$modifiers = document.querySelector('div.modifiers');
        this.$operations = document.querySelector('div.operations');
        this.$total = document.querySelector('#total');
        this.initializeComponents();
    }
```
이렇게 클래스에서 사용되는 멤버변수를 미리 정의해주면 이 클래스가 어떤 멤버변수를 사용하는지 한 눈에 파악할 수 있어서 좋답니다!

그리고 initState를 state에 바로 할당할 경우 레퍼런스를 공유하기 때문에 예상하는 것 처럼 동작하지 않을 확률도 있답니다!

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,103 @@
+        this.setState(initialState);
+    }
+
+    initializeComponents() {
+        new Digits({
+            $target: this.$digits,
+            updateOperand: this.updateOperand.bind(this),
+        });
+
+        new Modifiers({
+            $target: this.$modifiers,
+            resetState: this.resetState.bind(this),
+        });
+
+        new Operations({
+            $target: this.$operations,
+            setOperation: this.setOperation.bind(this),
+        });
+    }
```
라이프 사이클을 만들어주니 더 코드의 흐름을 이해하기가 수월하네요 👍

JunilHwang commented on src/js/components/Calculator.js:
```diff
@@ -0,0 +1,103 @@
+    updateOperand(digit) {
+        const { state } = this;
+        const { operation } = state;
+        let { leftOperand, rightOperand } = state;
+
+        if (!operation) {
+            leftOperand = formatOperand(leftOperand, digit);
+        } else {
+            rightOperand = formatOperand(rightOperand, digit);
+        }
+
+        const total = operation
+            ? leftOperand + operation + rightOperand
+            : leftOperand;
```
한 층 더 고민해볼 수 있는 부분은, 이런 로직을 View에서 처리해주는게 맞는걸까? 하는 것입니다.
일종의 자료 구조(?)가 존재하고, 해당 자료구조를 통해서 연산을 수행하는 중입니다.
이 부분을 추출하여 관리할 수 있지 않을까요?

JunilHwang reviewed (approved):
안녕하세요 태영님! 피드백 반영 빠르게 잘 해주셨네요 👍 
이번에도 몇 가지 피드백을 남기긴 했으나, 다음 미션을 진행하실 때 참고해보시면 될 것 같습니다 😁
계산기 미션은 여기서 마무리하도록 하겠습니다!

다음 미션인 로또 미션의 경우 Domain 이 더 명확한 미션입니다. 즉, 로또 라는 개념 자체에 대해서 이해가 필요한거죠 ㅎㅎ

따라서 View와 Domain을 더 철저하게 구분하고 관리할 수 있도록 신경쓰면서 미션을 진행해보시면 좋답니다!
지금 보다 더 철저하게 오직 렌더링을 수행하는 부분과 렌더링 이외의 부분을 구분하는거죠!

View와 Domain은 변화에 대한 목적이 다릅니다
View는 콘솔이 될 수도 있고, 네이티브 앱이 될 수도 있습니다.
Domain은 어디에서든 사용될 수 있습니다.

Domain을 일종의 외부 패키지라고 생각하면서 작성해보세요!

짧은 시간이었지만 부디 조금이나마 도움이 되었길 바랍니다 🙏
수고하셨습니다 🙇‍♂️
